### PR TITLE
Add predicates for find command

### LINKED ISSUE #119: As a Hall admin managing room, I want to search for students in a level, so that I can find out who stays on what level

### LINKED ISSUE #120: As a Hall admin managing rooms, I want to search for students in a room, to quickly find out who is staying in which room.

### LINKED ISSUE #121: As a hall admin managing residents, I want to filter residents from gender, so that I can find out what the gender ratio of the hall population is

### LINKED ISSUE #122: As a Hall admin finding residents, I want to search for students using their matriculation number, so that I can quickly identify students from their matriculation number

### LINKED ISSUE #123: As a hall admin, I want to quickly find residents based on their characteristics such as gender, blocks and levels, so that I can quickly look up students info easily.
Related issues #122 , #121, #120 , #119 , #118

### PR DESCRIPTION
**Pull request by athuyaoo**
Closes issues #119, #120, #121, #122, #123

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 1]
```diff
@@ -20,12 +24,16 @@
...
             + "Parameters: "
-            + "[" + PREFIX_NAME + "KEYWORDS [MORE KEYWORDS] " + "]"
-            + "[" + PREFIX_BLOCK + "BLOCK " + "]"
+            + "[" + PREFIX_NAME + "KEYWORDS [MORE KEYWORDS]" + "] "
+            + "[" + PREFIX_BLOCK + "BLOCK" + "] "
+            + "[" + PREFIX_FLOOR + "FLOOR" + "]"
+            + "[" + PREFIX_ROOM_NUMBER + "ROOM_NUMBER" + "] "
+            + "[" + PREFIX_MATRICULATION_NUMBER + "MATRICULATION_NUMBER" + "] "
+            + "[" + PREFIX_GENDER + "GENDER" + "] "
             + "[" + PREFIX_STUDENT_GROUP + "STUDENT_GROUP " + "]"
```
Comment (jieffyy) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:37]: Sorry this isn't exactly your issue, but actually should it be `"[" + PREFIX_STUDENT_GROUP + "STUDENT_GROUP " + "]..."` instead? Reason being the user can actually do `find s/dance s/basketball`.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -29,7 +40,7 @@
 public class ParserUtil {
 
     public static final String MESSAGE_INVALID_INDEX = "Index is not a non-zero unsigned integer.";
-
+    private static final Logger logger = LogsCenter.getLogger(ParserUtil.class);
```
Comment (jieffyy) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:43]: Neat!

File: src/test/java/seedu/address/model/person/MatriculationNumberMatchPredicateTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        // null -> returns false
+        assertFalse(firstPredicate.equals(null));
+
+        // different person -> returns false
+        assertFalse(firstPredicate.equals(secondPredicate));
+    }
+
+    @Test
+    public void test_nameContainsKeywords_returnsTrue() {
```
Comment (jieffyy) [inline src/test/java/seedu/address/model/person/MatriculationNumberMatchPredicateTest.java]: Is there a typo in the method name - should it be `test_matriculationNumber_returnsTrue()`?

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -244,4 +255,97 @@ public static Location parseLocation(String location) throws ParseException {
...
+            String roomNumber) throws ParseException {
+        logger.log(Level.INFO, "parsing RoomMatchesNumberPredicate");
+        requireNonNull(roomNumber);
+        String trimmedRoomNumber = roomNumber.trim();
+        if (!RoomMatchesNumberPredicate.isValidRoomNumber(trimmedRoomNumber)) {
+            throw new ParseException(RoomMatchesNumberPredicate.MESSAGE_CONSTRAINTS);
+        }
+        if (!RoomMatchesNumberPredicate.isValidRoomNumber(trimmedRoomNumber)) {
+            throw new ParseException(RoomMatchesNumberPredicate.MESSAGE_CONSTRAINTS);
+        }
```
Comment (jieffyy) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: Are these 2 the same?

**Review by jieffyy — COMMENTED** [review]
Almost LGTM. Some minor nits.

File: src/main/java/seedu/address/logic/commands/FindCommand.java (hunk 1 above)
Comment (athuyaoo) [reply src/main/java/seedu/address/logic/commands/FindCommand.java:37]: isn't it already that?

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 2 above)
Comment (athuyaoo) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:43]: Thanks!

File: src/test/java/seedu/address/model/person/MatriculationNumberMatchPredicateTest.java (hunk 3 above)
Comment (athuyaoo) [reply src/test/java/seedu/address/model/person/MatriculationNumberMatchPredicateTest.java]: Thanks

File: src/main/java/seedu/address/logic/commands/FindCommand.java (hunk 1 above)
Comment (jieffyy) [reply src/main/java/seedu/address/logic/commands/FindCommand.java:37]: Okay never mind, my bad.

**Review by jieffyy — APPROVED** [review]

**Pull request merged**